src/Vulkan/Silk.NET.Vulkan/Vk.cs

```diff
@@ -197,18 +225,69 @@ public bool IsDeviceExtensionPresent(Instance instance, string extension)
+            var prefix = device.Handle.ToString();
+            var prefix_sep = prefix + '|';
+            var fullKey = prefix_sep + extension;
+            var result = false;
+
+            // We place a devices handle into the hashset to indicate it has been previously loaded.
+            // We then prefix entries with '<Handle>|' which will never collide with '<Handle>' (which is an
+            // integer and so doesn't contain the pipe char '|').  This prevents doing two lookups, on cache hit and so
+            // is twice as fast as using a Dictionary<string, HashSet<string>> in best case (and significantly faster than
+            // Dictionary<string, List<string>>))
+            // 
+            // Note we use an upgradeable read lock, which is very fast.  We avoid a try...finally block on read
+            // as the operations are inheritently safe, the additional try...finally on write is mitigated
+            // as we also use it to dispose the memory allocation safely.  This gives us thread-safety at
+            // no real additional cost.
+            _cachedDeviceExtensionsLock.EnterUpgradeableReadLock();
+
+            // We check for the extension first to avoid 2 lookups
+            if (_cachedDeviceExtensions.Contains(fullKey))
             {
-                var extProperties = stackalloc ExtensionProperties[128];
-                _cachedDeviceExtensions.Add(device.Handle, list = Add(device, new List<string>(), extProperties));
+                // We found the extension
+                result = true;
             }
-            else
+            else if (!_cachedDeviceExtensions.Contains(prefix))
             {
-                list = _cachedDeviceExtensions[device.Handle];
-            }
+                // The lack of the device handle indicates we've not been previously initialised.  We now need a write lock.
+                _cachedDeviceExtensionsLock.EnterWriteLock();
+                GlobalMemory mem = null;
+                try
+                {
+                    var deviceExtPropertiesCount = 0u;
+
+                    // Get number of properties
+                    EnumerateDeviceExtensionProperties(device, (byte*) 0, &deviceExtPropertiesCount, null);
+
+                    // Initialise return structure
+                    mem = GlobalMemory.Allocate((int) deviceExtPropertiesCount * sizeof(ExtensionProperties));
+                    var props = (ExtensionProperties*) Unsafe.AsPointer(ref mem.GetPinnableReference());
 
-            return list.Contains(extension);
+                    // Get properties
+                    EnumerateDeviceExtensionProperties(device, (byte*) 0, &deviceExtPropertiesCount, props);
+                    for (int j = 0; j < deviceExtPropertiesCount; j++)
+                    {
+                        // Prefix the extension name
+                        var newKey = prefix_sep + Marshal.PtrToStringAnsi((IntPtr) props[j].ExtensionName);
```

Comment:
Preferably use SilkMarshal over Marshal to convert from ptr to string

Reply:
This was using the existing code.  I am unsure of how `SilkMarshal` works and whether it is equivalent to `Marshal` which would therefore require more testing.

Reply:
I would do this but there's also a lot of mixed-use of `Marshal` and `SilkMarshal` in `HelloTriangleApplications.cs`, and the rest of the codebase for that matter.  It doesn't strike me as a change relating to this PR to review all `Marshal` usages and update to `SilkMarshal`?  However, if this is required, there are 2 places in `VK.cs` and 1 in `XR.cs` which use `Marshal`, specifically `Marshal.PtrToStringAnsi`, and I assume can be converted to `SilkMarshal.PtrToString`.

Reply:
Doesn't really matter in this instance, but the Mono implementation of the Marshal methods are pretty inconsistent. Doesn't have to be now, but eventually we should switchover to SilkMarshal fully.

Reply:
Further to @Perksey this is for a later review.